src/Raven.Server/RavenServerStartup.cs

```diff
@@ -246,6 +239,35 @@ private async Task RequestHandler(HttpContext context)
             }
         }
 
+        private static void CheckMajorVersionClientServerAndUpdateException(HttpContext context, ref Exception e)
+        {
+            if (context.Request.Headers.TryGetValue(Constants.Headers.ClientVersion, out var versions) == false) 
+                return;
+            
+            var version = versions.ToString();
+            if (version.Length <= 0) 
+                return;
+            
+            if (version[0] == '3')
+            {
+                e = new ClientVersionMismatchException(
+                    $"RavenDB does not support interaction between Client API major version 3 and Server version {RavenVersionAttribute.Instance.MajorVersion} when major version does not match. " +
+                    $"Client: {version}. Server: {RavenVersionAttribute.Instance.AssemblyVersion}",
+                    e);
+                return;
+            }
+
+            var nVersion = char.GetNumericValue(version, 0);
+            if (nVersion > RavenVersionAttribute.Instance.MajorVersion)
+            {
+                e = new ClientVersionMismatchException(
```

Comment:
What will happen if we throw e.g. 'IndexDoesNotExistException' when Client version is 5.0.0.50?

Reply:
So we will get a warning about the version mismatch.

Reply:
Which is incorrect? Right?